File: leetcode/lc-us/cpp/1129.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>> &redEdges, vector<vector<int>> &blueEdges) {
        const int INF = 0x3f3f3f3f;
        const int RED = 0, BLUE = 1;
        vector<vector<int>> rg(n), bg(n);
        for (auto &e: redEdges) rg[e[0]].push_back(e[1]);
        for (auto &e: blueEdges) bg[e[0]].push_back(e[1]);
        int dist[n][2];
        memset(dist, 0x3f, sizeof dist);
        dist[0][0] = dist[0][1] = 0;
        queue<tuple<int, int, int>> q;
        q.push({0, RED, 0});
        q.push({0, BLUE, 0});
        while (!q.empty()) {
            auto[p, c, d] = q.front();
            q.pop();
            if (c == RED) {
                for (int ne: bg[p]) {
                    if (dist[ne][BLUE] > d + 1) {
                        dist[ne][BLUE] = d + 1;
                        q.push({ne, BLUE, d + 1});
                    }
                }
            } else {
                for (int ne: rg[p]) {
                    if (dist[ne][RED] > d + 1) {
                        dist[ne][RED] = d + 1;
                        q.push({ne, RED, d + 1});
                    }
                }
            }
        }
        vector<int> res(n);
        for (int i = 0; i < n; i++) {
            int t = min(dist[i][0], dist[i][1]);
            if (t == INF) res[i] = -1;
            else res[i] = t;
        }
        return res;
    }
};

File: leetcode/lc-us/py/1129.py
from collections import defaultdict

from typing import List


class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        g = defaultdict(list)
        for e in redEdges:
            g[e[0]].append((e[1], 0))
        for e in blueEdges:
            g[e[0]].append((e[1], 1))
        r, b = [0] + [-1] * (n - 1), [0] + [-1] * (n - 1)
        q = [(0, 0), (0, 1)]
        while q:
            t = q[0]
            q = q[1:]
            p, c = t
            for ne in g[p]:
                if c == 0:
                    if ne[1] == 1 and (b[ne[0]] == -1 or b[ne[0]] > r[p] + 1):
                        b[ne[0]] = r[p] + 1
                        q.append((ne[0], 1))
                else:
                    if ne[1] == 0 and (r[ne[0]] == -1 or r[ne[0]] > b[p] + 1):
                        r[ne[0]] = b[p] + 1
                        q.append((ne[0], 0))
        res = [-1] * n
        for i in range(n):
            if r[i] == b[i] == -1:
                continue
            elif r[i] == -1:
                res[i] = b[i]
            elif b[i] == -1:
                res[i] = r[i]
            else:
                res[i] = min(r[i], b[i])
        return res
